### PR TITLE
server: fix Setting wait_timeout to 0 doesn't take effect #23351

### LINKED ISSUE #23351: Setting `wait_timeout` to 0 doesn't take effect
## Bug Report

After setting `wait_timeout` to a non-zero value and then setting it back to 0, the connection will still timeout.

```go
func (p *packetIO) readOnePacket() ([]byte, error) {
	var header [4]byte
	if p.readTimeout > 0 {
		if err := p.bufReadConn.SetReadDeadline(time.Now().Add(p.readTimeout)); err != nil {
			return nil, err
		}
	}
...
```

In this function, setting `wait_timeout` to 0 won't go into `SetReadDeadline`, so it doesn't take effect.

### 1. Minimal reproduce step (Required)

```java
            Statement stmt = conn.createStatement();
            stmt.execute("set @@wait_timeout=1");
            stmt.execute("set @@wait_timeout=0");
            Thread.sleep(3000);
            ResultSet rs = stmt.executeQuery("select 1");
```

### 2. What did you expect to see? (Required)

The query is executed successfully.

### 3. What did you see instead (Required)

It reports `CommunicationsException`.

### 4. What is your TiDB version? (Required)

v3.0.17

### PR DESCRIPTION
**Pull request by Howie59**
### What problem does this PR solve?

Issue Number: close #23351 

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)
```sql
set @@wait_timeout=10;
set @@wait_timeout=0;
<!--wait 10 seconds-->
show variables like "%wait_timeout%";
select 1;
<!--got no reconnect or error-->
```
### Release note 
- No release note

### CODE REVIEW

File: server/packetio.go [hunk 1]
```diff
@@ -96,8 +100,12 @@ func (p *packetIO) readOnePacket() ([]byte, error) {
 	length := int(uint32(header[0]) | uint32(header[1])<<8 | uint32(header[2])<<16)
 
 	data := make([]byte, length)
-	if p.readTimeout > 0 {
-		if err := p.bufReadConn.SetReadDeadline(time.Now().Add(p.readTimeout)); err != nil {
+	if p.readTimeout >= 0 {
+		readTimeout := time.Now().Add(p.readTimeout)
+		if p.readTimeout == 0 {
```
Comment (mmyj) [inline server/packetio.go:105]: L85 has reset the deadline,  should we need to reset again here?

**Comment by mmyj** [comment]
Could we move the reset logic to the `packetIO.readPacket()`, like following? 
image

**Comment by Howie59** [comment]
> Could we move the reset logic to the `packetIO.readPacket()`, like following?
> image

ok, i will modify this one

**Pull request merged**